# Hugging Face Tutorial Lab: `pipeline()`

## 1. Transformers, what can they do?

In [1]:
# pip install "transformers==4.38.2"
# pip install torch datasets evaluate "transformers[sentencepiece]" accelerate

import sys
print(sys.executable)

/Users/sherry/Nextcloud/Teaching Work/2026-2 Spring - Advanced AI and ML/LabNLP/.venv/bin/python


In [2]:
import torch
import transformers

print("Torch:", torch.__version__)
print("Transformers:", transformers.__version__)

Torch: 2.2.2
Transformers: 4.38.2


/Users/sherry/Nextcloud/Teaching Work/2026-2 Spring - Advanced AI and ML/LabNLP/.venv/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [3]:
from transformers import pipeline

classifier = pipeline("sentiment-analysis")

# Test 1: single sentence
result1 = classifier("I've been waiting for a Hugging Face course my whole life.")
print(result1)

# Test 2: multiple sentences
texts = [
    "I've been waiting for a Hugging Face course my whole life.",
    "I hate this so much!"
]

result2 = classifier(texts)
print(result2)

2026-05-19 23:25:01.791338: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
/Users/sherry/Nextcloud/Teaching Work/2026-2 Spring - Advanced AI and ML/LabNLP/.venv/lib/python3.10/site-packages/huggingface_hub/file_download.py:949: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


[{'label': 'POSITIVE', 'score': 0.9982948899269104}]
[{'label': 'POSITIVE', 'score': 0.9982948899269104}, {'label': 'NEGATIVE', 'score': 0.9994558691978455}]


In [4]:
#from transformers import pipeline

classifier = pipeline("zero-shot-classification")
classifier(
    "This is a course about the Transformers library",
    candidate_labels=["education", "politics", "business"],
)

No model was supplied, defaulted to facebook/bart-large-mnli and revision c626438 (https://huggingface.co/facebook/bart-large-mnli).
Using a pipeline without specifying a model name and revision in production is not recommended.
/Users/sherry/Nextcloud/Teaching Work/2026-2 Spring - Advanced AI and ML/LabNLP/.venv/lib/python3.10/site-packages/huggingface_hub/file_download.py:949: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


{'sequence': 'This is a course about the Transformers library',
 'labels': ['education', 'business', 'politics'],
 'scores': [0.8445985913276672, 0.11197441071271896, 0.04342696815729141]}

In [5]:
#from transformers import pipeline

generator = pipeline("text-generation")
generator("In this course, we will teach you how to")

No model was supplied, defaulted to openai-community/gpt2 and revision 6c0e608 (https://huggingface.co/openai-community/gpt2).
Using a pipeline without specifying a model name and revision in production is not recommended.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': "In this course, we will teach you how to write code using OCaml. Ocaml is a collaborative programming language designed to be used by programmers.\n\nHere's a short tutorial video by Justin M. Leung showing you how to"}]

In [6]:
#from transformers import pipeline

generator = pipeline("text-generation", model="distilgpt2")
generator(
    "In this course, we will teach you how to",
    max_length=30,
    num_return_sequences=2,
)

Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': 'In this course, we will teach you how to be creative and creative and how to master your skills when you need it:\n\n1. Find'},
 {'generated_text': 'In this course, we will teach you how to use our methods to create custom objects, to create an actual object, and use our methods to create'}]

In [7]:
#from transformers import pipeline

unmasker = pipeline("fill-mask")
unmasker("This course will teach you all about <mask> models.", top_k=2)

No model was supplied, defaulted to distilbert/distilroberta-base and revision ec58a5b (https://huggingface.co/distilbert/distilroberta-base).
Using a pipeline without specifying a model name and revision in production is not recommended.
Some weights of the model checkpoint at distilbert/distilroberta-base were not used when initializing RobertaForMaskedLM: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
- This IS expected if you are initializing RobertaForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


[{'score': 0.19619861245155334,
  'token': 30412,
  'token_str': ' mathematical',
  'sequence': 'This course will teach you all about mathematical models.'},
 {'score': 0.04052719846367836,
  'token': 38163,
  'token_str': ' computational',
  'sequence': 'This course will teach you all about computational models.'}]

In [8]:
#from transformers import pipeline
# NER means Named Entity Recognition
#Aggregate token-level predictions into word/entity-level predictions using a simple grouping rule.
ner = pipeline("ner", aggregation_strategy="simple")
ner("My name is Sylvain and I work at Hugging Face in Brooklyn.")

No model was supplied, defaulted to dbmdz/bert-large-cased-finetuned-conll03-english and revision f2482bf (https://huggingface.co/dbmdz/bert-large-cased-finetuned-conll03-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
Some weights of the model checkpoint at dbmdz/bert-large-cased-finetuned-conll03-english were not used when initializing BertForTokenClassification: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']
- This IS expected if you are initializing BertForTokenClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForTokenClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


[{'entity_group': 'PER',
  'score': 0.9981694,
  'word': 'Sylvain',
  'start': 11,
  'end': 18},
 {'entity_group': 'ORG',
  'score': 0.9796019,
  'word': 'Hugging Face',
  'start': 33,
  'end': 45},
 {'entity_group': 'LOC',
  'score': 0.9932106,
  'word': 'Brooklyn',
  'start': 49,
  'end': 57}]

In [9]:
#from transformers import pipeline

question_answerer = pipeline("question-answering")
question_answerer(
    question="Where do I work?",
    context="My name is Sylvain and I work at Hugging Face in Brooklyn",
)

No model was supplied, defaulted to distilbert/distilbert-base-cased-distilled-squad and revision 626af31 (https://huggingface.co/distilbert/distilbert-base-cased-distilled-squad).
Using a pipeline without specifying a model name and revision in production is not recommended.


{'score': 0.6949770450592041, 'start': 33, 'end': 45, 'answer': 'Hugging Face'}

In [10]:
#from transformers import pipeline

summarizer = pipeline("summarization")
summarizer(
    """
    America has changed dramatically during recent years. Not only has the number of 
    graduates in traditional engineering disciplines such as mechanical, civil, 
    electrical, chemical, and aeronautical engineering declined, but in most of 
    the premier American universities engineering curricula now concentrate on 
    and encourage largely the study of engineering science. As a result, there 
    are declining offerings in engineering subjects dealing with infrastructure, 
    the environment, and related issues, and greater concentration on high 
    technology subjects, largely supporting increasingly complex scientific 
    developments. While the latter is important, it should not be at the expense 
    of more traditional engineering.

    Rapidly developing economies such as China and India, as well as other 
    industrial countries in Europe and Asia, continue to encourage and advance 
    the teaching of engineering. Both China and India, respectively, graduate 
    six and eight times as many traditional engineers as does the United States. 
    Other industrial countries at minimum maintain their output, while America 
    suffers an increasingly serious decline in the number of engineering graduates 
    and a lack of well-educated engineers.
"""
)

No model was supplied, defaulted to sshleifer/distilbart-cnn-12-6 and revision a4f8f3e (https://huggingface.co/sshleifer/distilbart-cnn-12-6).
Using a pipeline without specifying a model name and revision in production is not recommended.


[{'summary_text': ' America has changed dramatically during recent years . The number of engineering graduates in the U.S. has declined in traditional engineering disciplines such as mechanical, civil,    electrical, chemical, and aeronautical engineering . Rapidly developing economies such as China and India continue to encourage and advance the teaching of engineering .'}]

In [11]:
#from transformers import pipeline

translator = pipeline("translation", model="Helsinki-NLP/opus-mt-fr-en")
translator("Ce cours est produit par Hugging Face.")

/Users/sherry/Nextcloud/Teaching Work/2026-2 Spring - Advanced AI and ML/LabNLP/.venv/lib/python3.10/site-packages/transformers/models/marian/tokenization_marian.py:197: UserWarning: Recommended: pip install sacremoses.
  warnings.warn("Recommended: pip install sacremoses.")


[{'translation_text': 'This course is produced by Hugging Face.'}]

## 2. What behind `pipeline()`? 
## - The simplest usage: `pipeline("sentiment-analysis")`

The Hugging Face `pipeline()` API hides three steps:

```text
1. Preprocessing
   raw text → tokens → token IDs → tensors

2. Model inference
   tensors → Transformer → classification head → logits

3. Postprocessing
   logits → softmax probabilities → labels and scores
```

The default sentiment-analysis checkpoint used in the Hugging Face tutorial is:

```python
distilbert-base-uncased-finetuned-sst-2-english
```

A **checkpoint** means a saved pretrained/fine-tuned model.



In [20]:
from transformers import pipeline

classifier = pipeline("sentiment-analysis")
# Test bias in the model: Muted may often be associated with negative sentiment

texts = [
    "I've been waiting for a Hugging Face course my whole life.",
    "I hate this so much!",
    "The movie was okay, but not very exciting.",
    "The movie has a muted visual style.",
]

results = classifier(texts)

for text, result in zip(texts, results):
    print("Text:", text)
    print("Result:", result)
    print("-" * 80)


No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.


Text: I've been waiting for a Hugging Face course my whole life.
Result: {'label': 'POSITIVE', 'score': 0.9982948899269104}
--------------------------------------------------------------------------------
Text: I hate this so much!
Result: {'label': 'NEGATIVE', 'score': 0.9994558691978455}
--------------------------------------------------------------------------------
Text: The movie was okay, but not very exciting.
Result: {'label': 'NEGATIVE', 'score': 0.9978721141815186}
--------------------------------------------------------------------------------
Text: The movie has a muted visual style.
Result: {'label': 'NEGATIVE', 'score': 0.9978736639022827}
--------------------------------------------------------------------------------


the pipeline is not just one simple function call. Internally, it performs this workflow:

```text
Raw text
   ↓
Tokenizer
   ↓
input_ids + attention_mask
   ↓
Transformer model
   ↓
logits
   ↓
softmax
   ↓
label + score
```

Next, we manually reproduce these steps.


### Step 1: Preprocessing - Tokenization

Transformer models cannot directly process raw text, so the tokenizer converts text into numerical inputs.
It usually does three things:

1. Split text into tokens.
2. Map tokens to token IDs.
3. Create extra inputs, such as `attention_mask`.

In [21]:
# Set checkpoint as a specific model
# You can set checkpoint to a different model that is fine-tuned on the same task, 
# or even a different task, to see how it performs.

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
print(checkpoint)

distilbert-base-uncased-finetuned-sst-2-english


In [33]:
# [CLS]=101(classification); [SEP]=102(end of sentence); [PAD]=0; [UNK]=100; [MASK]=103

from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained(checkpoint)

raw_inputs = [
    "I've been waiting for a Hugging Face course my whole life.",
    "I hate this so much!",
]

inputs = tokenizer(
    raw_inputs,
    padding=True,
    truncation=True,
    return_tensors="pt"
)
# padding: fill short sentence  
# truncation: cut long sentence (longer than 512 tokens) 
# pt: Return PyTorch tensors.
# print(inputs)


*What are `input_ids` and `attention_mask`?*

`input_ids` are token IDs. For example, a word or subword is mapped to an integer.
`attention_mask` tells the model which positions are real tokens and which positions are padding.

Usually:
```text
1 = real token
0 = padding token
```

*Notice special tokens*

For BERT-style models:
- `[CLS]` is often used as the sentence-level representation.
- `[SEP]` marks the end of a sentence.
- `[PAD]` is added to make sequences in the same batch have the same length.

Check tokenizer in: https://huggingface.co/spaces/Xenova/the-tokenizer-playground


In [36]:
print(inputs)
print("input_ids shape:", inputs["input_ids"].shape)
print("attention_mask shape:", inputs["attention_mask"].shape)

print("\ninput_ids:")
print(inputs["input_ids"])

print("\nattention_mask:")
print(inputs["attention_mask"])


{'input_ids': tensor([[  101,  1045,  1005,  2310,  2042,  3403,  2005,  1037, 17662,  2227,
          2607,  2026,  2878,  2166,  1012,   102],
        [  101,  1045,  5223,  2023,  2061,  2172,   999,   102,     0,     0,
             0,     0,     0,     0,     0,     0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0]])}
input_ids shape: torch.Size([2, 16])
attention_mask shape: torch.Size([2, 16])

input_ids:
tensor([[  101,  1045,  1005,  2310,  2042,  3403,  2005,  1037, 17662,  2227,
          2607,  2026,  2878,  2166,  1012,   102],
        [  101,  1045,  5223,  2023,  2061,  2172,   999,   102,     0,     0,
             0,     0,     0,     0,     0,     0]])

attention_mask:
tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0]])


In [37]:
#Look at tokens directly
for i, sentence_ids in enumerate(inputs["input_ids"]):
    tokens = tokenizer.convert_ids_to_tokens(sentence_ids)
    print(f"Sentence {i + 1}:")
    print(tokens)
    print()


Sentence 1:
['[CLS]', 'i', "'", 've', 'been', 'waiting', 'for', 'a', 'hugging', 'face', 'course', 'my', 'whole', 'life', '.', '[SEP]']

Sentence 2:
['[CLS]', 'i', 'hate', 'this', 'so', 'much', '!', '[SEP]', '[PAD]', '[PAD]', '[PAD]', '[PAD]', '[PAD]', '[PAD]', '[PAD]', '[PAD]']



### Step 2: Model inference
### *2.1 Pass inputs through the base model*

First, we use `AutoModel`.
`AutoModel` loads the base Transformer model **without the classification head**.
It outputs hidden states/features, not final sentiment labels.


In [38]:
from transformers import AutoModel

# Load a pretrained Transformer model from the checkpoint. 
# no training, but just loads saved weights
base_model = AutoModel.from_pretrained(checkpoint)

#Send the tokenized inputs through the model once. (forward pass)
#Same as:
# base_outputs = base_model(
#    input_ids=inputs["input_ids"],
#    attention_mask=inputs["attention_mask"])
base_outputs = base_model(**inputs)

#Show the shape of the model’s output.
print(base_outputs.last_hidden_state.shape)


/Users/sherry/Nextcloud/Teaching Work/2026-2 Spring - Advanced AI and ML/LabNLP/.venv/lib/python3.10/site-packages/huggingface_hub/file_download.py:949: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


torch.Size([2, 16, 768])


#### Understanding the shape

The output shape is usually:  [batch_size, sequence_length, hidden_size]

For this example, it is likely: [2, 16, 768] 
- `2`: two input sentences
- `16`: token length after padding
- `768`: hidden vector dimension for each token

This is not yet a classification result.

### *2.2 Use a model with a classification head*

AutoModel does choose the correct base model architecture from the checkpoint, but it only loads the base Transformer body. 

If you want sentiment classification, you need: 
```python
AutoModelForSequenceClassification
```
This adds a task-specific classification head on top of the base Transformer.


Here, let's use 
```python
AutoModel 
```
which choose the base model architecture from the checkpoint, and check what inside of pipeline() 


In [39]:
from transformers import AutoModelForSequenceClassification

classification_model = AutoModelForSequenceClassification.from_pretrained(checkpoint)

outputs = classification_model(**inputs)

print(outputs.logits)
print("logits shape:", outputs.logits.shape)


tensor([[-3.1071,  3.2654],
        [ 4.1692, -3.3464]], grad_fn=<AddmmBackward0>)
logits shape: torch.Size([2, 2])


*What are logits?*

`logits` are the raw, unnormalized scores produced by the final layer of the model.

They are **not probabilities**.

For sentiment analysis with two labels, each sentence gets two logits: 
[score_for_NEGATIVE, score_for_POSITIVE]

So the shape is: [batch_size, number_of_labels], 
e.g.[2, 2]: There are 2 sentences into a 2-class sentiment model.

output: {[-3.1071,  3.2654],[ 4.1692, -3.3464]} 
so, in the first sentence, "negative=-3.1071, positive=3.2654"

In [41]:
print("id2label mapping:")
print(classification_model.config.id2label)

print("\nlabel2id mapping:")
print(classification_model.config.label2id)

id2label mapping:
{0: 'NEGATIVE', 1: 'POSITIVE'}

label2id mapping:
{'NEGATIVE': 0, 'POSITIVE': 1}


### Step 3: Post Analysis - Convert logits to probabilities with softmax

The model output logits must be converted into probabilities.
We use:

```python
softmax(logits)
```
The largest logit usually becomes the largest probability.


In [42]:
import torch

probabilities = torch.nn.functional.softmax(outputs.logits, dim=-1)
print(probabilities)


tensor([[1.7051e-03, 9.9829e-01],
        [9.9946e-01, 5.4418e-04]], grad_fn=<SoftmaxBackward0>)


### Convert probabilities into labels
Now we manually reproduce the final output of the pipeline.

In [43]:
predicted_class_ids = torch.argmax(probabilities, dim=-1)

for i, text in enumerate(raw_inputs):
    class_id = predicted_class_ids[i].item()
    label = classification_model.config.id2label[class_id]
    score = probabilities[i][class_id].item()

    print("Text:", text)
    print("Predicted label:", label)
    print("Score:", score)
    print("-" * 80)


Text: I've been waiting for a Hugging Face course my whole life.
Predicted label: POSITIVE
Score: 0.9982948899269104
--------------------------------------------------------------------------------
Text: I hate this so much!
Predicted label: NEGATIVE
Score: 0.9994558691978455
--------------------------------------------------------------------------------


### Compare manual result with `pipeline()`

Now we check whether our manual process gives the same meaning as the high-level pipeline.


In [44]:
classifier = pipeline("sentiment-analysis", model=checkpoint, tokenizer=checkpoint)

pipeline_results = classifier(raw_inputs)

for text, result in zip(raw_inputs, pipeline_results):
    print("Text:", text)
    print("Pipeline result:", result)
    print("-" * 80)


Text: I've been waiting for a Hugging Face course my whole life.
Pipeline result: {'label': 'POSITIVE', 'score': 0.9982948899269104}
--------------------------------------------------------------------------------
Text: I hate this so much!
Pipeline result: {'label': 'NEGATIVE', 'score': 0.9994558691978455}
--------------------------------------------------------------------------------


### Try your own examples
Can you modify the following texts and observe the output.

In [47]:
student_texts = [
    "This course is difficult but very useful.",
    "The homework is confusing and frustrating.",
    "I am not sure whether I like this movie.",
    "The product is not bad at all.",
]

student_results = classifier(student_texts)

for text, result in zip(student_texts, student_results):
    print("Text:", text)
    print("Result:", result)
    print("-" * 80)


Text: This course is difficult but very useful.
Result: {'label': 'POSITIVE', 'score': 0.9954234957695007}
--------------------------------------------------------------------------------
Text: The homework is confusing and frustrating.
Result: {'label': 'NEGATIVE', 'score': 0.9991816878318787}
--------------------------------------------------------------------------------
Text: I am not sure whether I like this movie.
Result: {'label': 'NEGATIVE', 'score': 0.9955669045448303}
--------------------------------------------------------------------------------
Text: The product is not bad at all.
Result: {'label': 'POSITIVE', 'score': 0.998784601688385}
--------------------------------------------------------------------------------


### limitation: sentence meaning can be subtle

A sentiment model may struggle with:

- sarcasm
- mixed opinions
- negation
- domain-specific language
- long context

For example:

```text
This movie is so good that I almost fell asleep.
```

The surface words may look positive, but the actual meaning is negative/sarcastic.


In [48]:
tricky_texts = [
    "This movie is so good that I almost fell asleep.",
    "The phone is cheap, but surprisingly reliable.",
    "I expected to hate it, but I actually loved it.",
    "Not bad at all.",
]

tricky_results = classifier(tricky_texts)

for text, result in zip(tricky_texts, tricky_results):
    print("Text:", text)
    print("Result:", result)
    print("-" * 80)


Text: This movie is so good that I almost fell asleep.
Result: {'label': 'POSITIVE', 'score': 0.9998125433921814}
--------------------------------------------------------------------------------
Text: The phone is cheap, but surprisingly reliable.
Result: {'label': 'POSITIVE', 'score': 0.9980798959732056}
--------------------------------------------------------------------------------
Text: I expected to hate it, but I actually loved it.
Result: {'label': 'POSITIVE', 'score': 0.9998137354850769}
--------------------------------------------------------------------------------
Text: Not bad at all.
Result: {'label': 'POSITIVE', 'score': 0.99928218126297}
--------------------------------------------------------------------------------


## 14. Optional: Use GPU if available

If PyTorch detects a CUDA GPU, you can ask the pipeline to use it.

```python
pipeline(..., device=0)
```

If no GPU is available, use CPU.


In [49]:
import torch
from transformers import pipeline

device = 0 if torch.cuda.is_available() else -1

print("CUDA available:", torch.cuda.is_available())
print("Pipeline device:", device)

gpu_classifier = pipeline(
    "sentiment-analysis",
    model=checkpoint,
    tokenizer=checkpoint,
    device=device
)

gpu_classifier("This is a quick GPU/CPU test.")


CUDA available: False
Pipeline device: -1


[{'label': 'NEGATIVE', 'score': 0.9968799352645874}]

## Try yourself
Can you complete the following tasks? 
1. Choose five English sentences.
2. Run them through `pipeline("sentiment-analysis")`.
3. Manually tokenize them with `AutoTokenizer`.
4. Pass them through `AutoModelForSequenceClassification`.
5. Print logits.
6. Apply softmax.
7. Compare the manual result with the pipeline result.

Expected conclusion:
> The pipeline result is produced by tokenization, model inference, and postprocessing.


In [50]:
# Student exercise starter code

my_texts = [
    # Add your own sentences here
]

# 1. Use pipeline
# my_pipeline_results = classifier(my_texts)

# 2. Tokenize
# my_inputs = tokenizer(my_texts, padding=True, truncation=True, return_tensors="pt")

# 3. Model inference
# my_outputs = classification_model(**my_inputs)

# 4. Softmax
# my_probs = torch.nn.functional.softmax(my_outputs.logits, dim=-1)

# 5. Print results
# print(my_outputs.logits)
# print(my_probs)
